beluga/include/beluga/algorithm/unscented_transform.hpp

```diff
@@ -0,0 +1,107 @@
+ * See https://en.wikipedia.org/wiki/Unscented_transform for more information.
+ *
+ * \tparam MeanDerived Concrete Eigen dense type used for the mean.
+ * \tparam CovarianceDerived Concrete Eigen dense type used for the covariance.
+ * \tparam TransferFn Callable that maps a vector from input space to output space. Possibly non-linear.
+ * \param mean Mean in the input space.
+ * \param covariance Covariance in the input space.
+ * \param transfer_fn Callable so that
+ * \param kappa Parameter used for sigma points selection. See https://arxiv.org/pdf/2104.01958 Equation (5). A sensible
+ * default will be used if not provided.
+ * \return A pair containing (mean, estimate) for the output space.
+ */
+template <typename MeanDerived, typename CovarianceDerived, typename TransferFn>
+auto unscented_transform(
+    const Eigen::MatrixBase<MeanDerived>& mean,
+    const Eigen::MatrixBase<CovarianceDerived>& covariance,
+    TransferFn&& transfer_fn,
+    std::optional<typename MeanDerived::Scalar> kappa = std::nullopt) {
+  using Scalar = typename MeanDerived::Scalar;
+  static_assert(
+      std::is_same_v<typename MeanDerived::Scalar, typename CovarianceDerived::Scalar>,
+      "Mean and covariance scalar types differ.");
+  static_assert(std::is_floating_point_v<Scalar>);
+  static_assert(MeanDerived::ColsAtCompileTime == 1, "Mean should be a column vector.");
+
+  constexpr int kNin = MeanDerived::RowsAtCompileTime;
+  const double k = kappa.value_or(std::max(kNin - 3, 0));
+  assert(k >= 0);
+  const Scalar w0 = static_cast<Scalar>(k) / static_cast<Scalar>(kNin + k);
+  std::vector<Eigen::Vector<Scalar, kNin>> sigma_points;
+  std::vector<Scalar> weights;
+  sigma_points.reserve(2 * kNin + 1);
+  weights.reserve(2 * kNin + 1);
+  sigma_points.emplace_back(mean);
+  weights.emplace_back(w0);
+  const Scalar wn = 1. / (2. * (kNin + k));
+
+  // Compute the Cholesky decomposition of the covariance matrix
+  const Eigen::Matrix<Scalar, kNin, kNin> chol_cov = covariance.llt().matrixL();
+
+  for (int i = 0; i < kNin; ++i) {
+    sigma_points.emplace_back(mean + static_cast<Scalar>(std::sqrt(kNin + k)) * chol_cov.col(i));
+    weights.emplace_back(wn);
+    sigma_points.emplace_back(mean - static_cast<Scalar>(std::sqrt(kNin + k)) * chol_cov.col(i));
+    weights.emplace_back(wn);
+  }
+
+  auto transformed_sigma_points = sigma_points | ranges::views::transform(transfer_fn);
+
+  constexpr int kNout = decltype(transformed_sigma_points.front())::RowsAtCompileTime;
+  static_assert(
+      decltype(transformed_sigma_points.front())::ColsAtCompileTime == 1, "Output mean should be a column vector");
+
+  Eigen::Vector<Scalar, kNout> out_mean = Eigen::Vector<Scalar, kNout>::Zero();
+  Eigen::Matrix<Scalar, kNout, kNout> out_cov = Eigen::Matrix<Scalar, kNout, kNout>::Zero();
+
+  for (const auto& [vector, weight] : ranges::views::zip(transformed_sigma_points, weights)) {
+    out_mean += (vector * weight).eval();
+  }
```

Comment:
@serraramiro1 you may want to consider `ranges::numeric::accumulate` for both loops.